packages/contracts/contracts/SmartInvoice.sol

```diff
@@ -133,9 +138,17 @@ contract SmartInvoice is
         details = _details;
         wrappedNativeToken = _wrappedNativeToken;
 
+        if (_requireVerification) emit Verified(_client, address(this));
+
         emit Register(_client, _provider, amounts);
     }
 
+    // Client verifies address before deposits
+    function verify() public {
+        require(msg.sender == client, "!client");
+        emit Verified(client, address(this));
+    }
```

Comment:
Is this a function that is only used on the frontend? I don't see this used anywhere in the contract so this should be either a view function that returns a boolean (for frontend) or implemented as is in the receive() and/or withdraw() functions.

Reply:
the function just emits a verified event to ensure the client has access to the client address. I can change the visibility to external though (changed it to public during a testing phase.

Reply:
I don't think view functions can emit Events. We can't actually prevent users from sending ERC-20s to this contract, so there is no need to store extra state to check against. We're just using an Event so the subgraph picks it up and can prompt users accordingly.

Reply:
Oh because of the subgraph gotcha. Yeah events are only emitted in tx functions, not view functions.

So the only drawback I still see with though is that wouldn't it still require the client to pay an additional gas fee simply to see if they're verified? That's a worse UX for something that can be possibly worked around with a view function instead.

Reply:
Can you describe this view function? I'm not sure I follow. What we're trying to do here is make sure that the client is able to perform transactions (eventually be able to release or lock) before depositing. We also need to store that this has taken place, and we're just putting it in the logs since it is a UI only feature. If gas is a concern (on mainnet, etc) the creator should use `false` for `requireVerification` and triple check offline.

Reply:
If this was a view function, it would look like this:
```
function verify() public view returns (bool) {
     return msg.sender == client;
}
```

I understand the purpose, but in my opinion, this seems like it's currently optimizing for the subgraph rather than for the user. I don't think a person should pay a transaction just to confirm they are the client address, unless it also served to log that the client agreed to the terms of the invoice. What I mean by that is the client signs a message of the terms (maybe details or just the contract address) with their wallet and makes a tx to the contract to verify the signature. There is an ECDSA library that can be used for verifying the public key and message signed. Similar to how some NFT projects use for whitelists.

For example:
```
function verify(bytes signature) external {
     address signer = ECDSA.recover(bytes32(details), signature);
     require(signer == client && signer == msg.sender, "!client");
     emit Verified(...)
}
```

Reply:
And also, this isn't to see they are verified, they will see that by querying the subgraph to see that the client address has verified meaning that depositing should be safe.

Reply:
I think the result of this conversation is that we should have a view function that allows the client to fund the wallet directly from the client address without verifying.

Reply:
I was about to merge but realized the above potentially changes this PR.

If the agreed upon set-up now is that only the client account can deposit until it 'verifies' and enables non-client accounts to deposit, then the provider no longer has control over that option. Which should mean that the provider-supplied requireVerification param is no longer required?

Reply:
What automatic getter do you mean?

Hmm it would make it easier removing the param if that's our setup by default. Would save a little gas too for the deployer.

Reply:
I'm ok forcing verification if we want to go that way. Initially I was thinking it would default that way and we would have an optional "skipVerification" that would be if the provider knew everything was fine (repeat client, etc). I'll leave it to you to decide the best option here, since a default argument isn't really a thing in solidity.